client/main.cpp

```diff
@@ -61,51 +53,57 @@ int main(int argc, char **argv) {
-        // separate io thread so that getline doesn't block our websocket pingpong
-        std::string line;
-        while (std::getline(instream, line)) {
-            if (line.empty())
-                continue;            
-            if (line == "exit" || line == "quit")
-                break;
-            std::unique_lock<std::mutex> lock(processor_mutex);
-            processor_cv.wait(lock, [&processed] { return processed; });
-            auto earliest = client_session.earliest();
-            auto now = std::chrono::high_resolution_clock::now();
-            if (earliest > now) {
-                // if we do auto its nano seconds, fine
-                auto delay = earliest - now;
-                std::cout << "cooldown " << delay.count() << "ns" << std::endl;
-                std::this_thread::sleep_for(delay);
-            }
-            processed = false;
-            try {
-                // post the work to ioc
-                // we do not do any network io here
-                ioc.post([&client_session, &processor_mutex, &processor_cv, &processed, line]() {
-                    std::unique_lock<std::mutex> lock(processor_mutex);
-                    client_session.handle_cmd(line);
-                    processed = true;
-                    lock.unlock();
-                    processor_cv.notify_one();
-                });
-            } catch (std::exception &e) {
-                std::cout << e.what() << std::endl;
-                break;
-            }
+    boost::asio::io_context ioc{};
+    auto work_guard = boost::asio::make_work_guard(ioc);
+
+    std::thread io_worker([&ioc] { ioc.run(); });
+
+    auto session_ptr = std::make_shared<nibaclient::client_session>(host, port, ioc, ctx);
+    // This needs to be a shared_ptr to use shared_from_this(), which makes our coroutine
+    // safe to access all members at all times.
+    // We can't rely on destructor for cleanup as a long running coroutine might not be done yet.
+    // Another way would be that the destructor should wait for futures that the coroutine
+    // would set at the end its execution.
+    session_ptr->start();
+    // Use start so we can fetch the first line of input while being setup
+
+    std::string line{};
+
+    std::getline(instream, line);     // Read line while session is starting...
+    session_ptr->block_until_ready(); // Block until session is actually ready
+
+    for (; instream;) {
+        if (line.empty())
+            continue;
+        if (line == "exit" || line == "quit")
+            break;
+        auto earliest = session_ptr->earliest();
+        auto now = std::chrono::high_resolution_clock::now();
+        if (earliest > now) {
+            // If we do auto its nano seconds, fine
+            auto delay = earliest - now;
+            std::cout << "cooldown " << delay.count() << "ns" << std::endl;
+            std::this_thread::sleep_for(delay);
```

Comment:
Is this delay only for fight simulation at the client side now? In future, we may need such a delay for other commands.

Reply:
Its on both end, server will error out requests that are being sent too fast, so we sleep on the client to prevent it from sending it too fast. Yes we need it for other requests.